### PR TITLE
feat: use ruff for lint & format

### LINKED ISSUE #5066: [enhancement]: use ruff as the one true linter/formatter
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Contact Details

_No response_

### What should this feature add?

[Ruff](https://github.com/astral-sh/ruff) is a new, very fast python linter/formatter. It is a drop-in replacement for `flake8`, `isort` and `black`, which is exactly what we use.

Pydantic has just moved to `ruff`, so I think it's good enough for us.

### Alternatives

_No response_

### Additional Content

_No response_

### PR DESCRIPTION
**Pull request by psychedelicious**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [x] Bug Fix
- [x] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Description

This PR introduces [`ruff`](https://github.com/astral-sh/ruff) as the only linter and formatter needed for the project. It is really fast. Like, alarmingly fast.

It is a drop-in replacement for flake8, isort, black, and much more. I've configured it similarly to our existing config. 

Note: we had enabled a number of flake8 plugins but didn't have the packages themselves installed, so they did nothing. Ruff used the existing config, and found a good number of changes needed to adhere to those flake8 plugins. I've resolved all violations.

### Code changes

- many [flake8-comprehensions](https://docs.astral.sh/ruff/rules/#flake8-comprehensions-c4) violations, almost all auto-fixed
- a good handful of [flake8-bugbear](https://docs.astral.sh/ruff/rules/#flake8-bugbear-b) violations
- handful of [pycodestyle](https://docs.astral.sh/ruff/rules/#pycodestyle-e-w) violations
- some formatting

### Developer Experience

[Ruff integrates with most editors](https://docs.astral.sh/ruff/integrations/):
- Official VSCode extension
- `ruff-lsp` python package allows it to integrate with any LSP-capable editor (vim, emacs, etc)
- Can be configured as an external tool in PyCharm

### Github Actions

I've updated the `style-checks` action to use ruff, and deleted the `pyflakes` action.

## Related Tickets & Documents

- Closes #5066 

## QA Instructions, Screenshots, Recordings

Have a poke around, and run the app. There were some logic changes but it was all pretty straightforward.

~~Not sure how to best test the changed github action.~~ Looks like it just used the action from this PR, that's kinda unexpected but OK.

## Added/updated tests?

- [ ] Yes
- [ ] No : _please replace this line with details on why tests
      have not been included_

### CODE REVIEW

**Review by lstein — APPROVED** [review]
Just as I was getting used to flake8, black, isort and mypy!

**Review by brandonrising — APPROVED** [review]
lgtm!

File: invokeai/backend/util/hotfixes.py [hunk 1]
```diff
@@ -732,7 +732,9 @@ def forward(
 
         controlnet_down_block_res_samples = ()
 
-        for down_block_res_sample, controlnet_block in zip(down_block_res_samples, self.controlnet_down_blocks):
+        for down_block_res_sample, controlnet_block in zip(
+            down_block_res_samples, self.controlnet_down_blocks, strict=True
+        ):
```
Comment (psychedelicious) [inline invokeai/backend/util/hotfixes.py:737]: When `zip` has receives collections of different lengths, it silently truncates the collections to the length of the shortest collection.

This is likely not intended, so `ruff` wants us to always specify the `strict` arg. 

If `strict=True`, and the collections are of different length, an error is raised. If `strict=False`, you get the silent truncation behaviour.

File: invokeai/backend/util/hotfixes.py [hunk 2]
```diff
@@ -745,7 +747,9 @@ def forward(
             scales = torch.logspace(-1, 0, len(down_block_res_samples) + 1, device=sample.device)  # 0.1 to 1.0
 
             scales = scales * conditioning_scale
-            down_block_res_samples = [sample * scale for sample, scale in zip(down_block_res_samples, scales)]
+            down_block_res_samples = [
+                sample * scale for sample, scale in zip(down_block_res_samples, scales, strict=True)
+            ]
```
Comment (psychedelicious) [inline invokeai/backend/util/hotfixes.py:752]: When `zip` has receives collections of different lengths, it silently truncates the collections to the length of the shortest collection.

This is likely not intended, so `ruff` wants us to always specify the `strict` arg. 

If `strict=True`, and the collections are of different length, an error is raised. If `strict=False`, you get the silent truncation behaviour.

**Review by RyanJDick — APPROVED** [review]

File: invokeai/backend/ip_adapter/resampler.py [hunk 3]
```diff
@@ -56,7 +56,7 @@ def forward(self, x, latents):
         x = self.norm1(x)
         latents = self.norm2(latents)
 
-        b, l, _ = latents.shape
+        b, L, _ = latents.shape
```
[the remark this answers is not in the document]
Comment (psychedelicious) [reply invokeai/backend/ip_adapter/resampler.py:59]: `l` is an ambiguous - see [PEP8 - Names to Avoid](https://peps.python.org/pep-0008/#names-to-avoid)

**Comment by psychedelicious** [comment]
I ran `ruff check .` and `ruff format .` on this branch after merging main in. It was so fast that I thought it mustn't have done anything. Running with `-v` this makes sense:

```sh
... list of files it checked
[2023-11-13][07:55:14][ruff_cli::commands::check][DEBUG] Identified files to lint in: 51.039562ms
...
[2023-11-13][07:55:14][ruff_cli::commands::check][DEBUG] Checked 263 files in: 866.562µs
```

```sh
[2023-11-13][07:56:24][ruff_cli::commands::format][DEBUG] Formatted 262 files in 13.94ms
```

O_O

**Pull request merged**